framework/wazuh/core/indexer/bulk.py

```diff
@@ -0,0 +1,199 @@
+from typing import Optional, Dict, Any, List, Protocol, Coroutine
+from enum import Enum
+
+from opensearchpy import AsyncOpenSearch
+
+
+class BulkAction(str, Enum):
+    """Enum representing bulk actions for OpenSearch.
+
+    Attributes
+    ----------
+    CREATE : str
+        Action to create a new document.
+    CREATE_OR_UPDATE : str
+        Action to create a new document or update an existing one.
+    UPDATE : str
+        Action to update an existing document.
+    DELETE : str
+        Action to delete an existing document.
+    """
+    CREATE = "create"
+    CREATE_OR_UPDATE = "index"
+    UPDATE = "update"
+    DELETE = "delete"
+
+    @classmethod
+    def exists(cls, action: str) -> bool:
+        try:
+            cls(action)
+        except ValueError:
+            return False
+        return True
+
+
+class BulkMetadata:
+    """Class to handle metadata for a bulk action in OpenSearch.
+
+    Parameters
+    ----------
+    index : str
+        Name of the index in OpenSearch.
+    doc_id : Optional[str]
+        ID of the document in OpenSearch. Can be None for certain actions.
+    action : BulkAction
+        Type of bulk action to perform.
+    """
+    def __init__(self, index: str, doc_id: Optional[str], action: BulkAction):
+        self.index = index
+        self.doc_id = doc_id
+        self.action = action
+
+    def decode(self) -> Dict[str, Dict[str, str]]:
+        """Decode metadata into a dictionary format for the OpenSearch bulk API.
+
+        Returns
+        -------
+        Dict[str, Dict[str, str]]
+            Metadata in a dictionary format.
+        """
+        return {str(self.action.value): {'_index': self.index, '_id': self.doc_id}}
+
+
+class BulkDoc:
+    """Class to represent a bulk document action for OpenSearch.
```

Comment:
```suggestion
    """Class to represent a bulk document action for the Indexer.
```

Reply:
Changed in f3067e1c80a1254db438f9ef7b4bc8535cd89e81.